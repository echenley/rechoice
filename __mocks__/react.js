const React = require.requireActual('react')

React.Fragment = 'Fragment'

module.exports = React
