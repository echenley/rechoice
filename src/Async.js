// @flow

import * as React from 'react'
import type {
  LoadItems,
  LoadItemsData,
  LoadItemsCallback,
  Item,
  LoadingState,
} from './types'

type State = {
  items: Array<Item>,
  loadingState: LoadingState,
}

type AsyncProps = State & {
  onChange: *,
  onInputChange?: (inputValue: string) => mixed,
}

type Props = {
  children: (props: AsyncProps) => React.Element<any>,
  loadItems: LoadItems,
  minimumInput: number,
  onChange: *,
  onFocus?: () => mixed,
  onInputChange?: (inputValue: string) => mixed,
}

class Async extends React.Component<Props, State> {
  static defaultProps = {
    minimumInput: 0,
  }

  state = {
    items: [],
    loadingState: 'success',
  }

  constructor(...args: Array<Object>) {
    super(...args)
    const self: Object = this
    self.onChange = this.onChange.bind(this)
  }

  componentDidMount() {
    this._isMounted = true
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  _isMounted: boolean = false
  mostRecentCallback: ?LoadItemsCallback = null

  loadItems = (inputValue: string) => {
    const { loadItems, minimumInput } = this.props

    if (inputValue.length < minimumInput) {
      this.setState({
        loadingState: 'success',
        items: [],
      })

      return null
    }

    const callback = (err, data) => {
      if (callback !== this.mostRecentCallback) {
        return
      }

      if (this._isMounted) {
        this.setState({
          loadingState: err ? 'error' : 'success',
          items: (data && data.items) || [],
        })
      }
    }

    this.mostRecentCallback = callback

    this.setState({
      loadingState: 'loading',
    })

    const promise = loadItems(inputValue, callback)

    if (promise) {
      return promise.then(
        (data: LoadItemsData) => callback(null, data),
        (error: Object) => callback(error)
      )
    }

    return null
  }

  async onChange(value: *) {
    const promise = Promise.resolve(this.props.onChange(value))

    if (!promise || !promise.then) {
      return
    }

    this.setState({ loadingState: 'loading' })

    await promise.catch(() => {})

    if (this._isMounted) {
      this.setState({ loadingState: 'success' })
    }
  }

  onFocus = (...args: Array<any>) => {
    if (this.props.onFocus) {
      this.props.onFocus(...args)
    }

    if (this.mostRecentCallback === null) {
      this.loadItems('')
    }
  }

  onInputChange = (inputValue: string) => {
    const { onInputChange } = this.props

    if (onInputChange) {
      onInputChange(inputValue)
    }

    return this.loadItems(inputValue)
  }

  render() {
    const { children } = this.props
    const { items, loadingState } = this.state

    return children({
      items,
      loadingState,
      onChange: this.onChange,
      onFocus: this.onFocus,
      onInputChange: this.onInputChange,
    })
  }
}

export default Async
