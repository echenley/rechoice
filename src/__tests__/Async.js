import * as React from 'react'
import { shallow } from 'enzyme'
import Async from '../Async'

jest.useFakeTimers()

const renderComponent = props => {
  return shallow(
    <Async
      // eslint-disable-next-line react/no-children-prop
      children={asyncProps => <div>{JSON.stringify(asyncProps)}</div>}
      loadItems={() => Promise.resolve({ items: [{ id: 1 }] })}
      minimumInput={0}
      {...props}
    />
  )
}

describe('Async', () => {
  it('renders function children prop', () => {
    expect(renderComponent().getElement()).toMatchSnapshot()
  })

  describe('loadItems', () => {
    it('sets state correctly if minimumInput is not met', () => {
      const children = jest.fn(() => <div />)

      const component = renderComponent({
        children,
        minimumInput: 3,
      })

      component.setState({
        items: [{ id: 1 }, { id: 2 }],
      })

      children.mock.calls[0][0].onInputChange('ab')

      expect(component.state('loadingState')).toBe('success')
      expect(component.state('items')).toEqual([])
    })

    it('calls loadItems (promise version) and updates items and loadingState', async () => {
      const children = jest.fn(() => <div />)

      const component = renderComponent({
        children,
      })

      component.setState({
        items: [{ id: 1 }, { id: 2 }],
      })

      const promise = children.mock.calls[0][0].onInputChange('test')

      expect(component.state('loadingState')).toBe('loading')

      await promise

      expect(component.state('loadingState')).toBe('success')
      expect(component.state('items')).toEqual([{ id: 1 }])
    })

    it('calls loadItems (callback version) and updates items and loadingState', () => {
      return new Promise((resolve, reject) => {
        const children = jest.fn(() => <div />)

        const component = renderComponent({
          children,
          loadItems: (input, cb) => {
            setTimeout(() => {
              cb(null, { items: [{ id: 1 }] })

              component.update()

              try {
                expect(component.state('loadingState')).toBe('success')
                expect(component.state('items')).toEqual([{ id: 1 }])
              } catch (err) {
                return reject(err)
              }

              return resolve()
            }, 0)
          },
        })

        component.setState({
          items: [{ id: 1 }, { id: 2 }],
        })

        children.mock.calls[0][0].onInputChange('test')

        jest.runAllTimers()

        try {
          expect(component.state('loadingState')).toBe('loading')
        } catch (err) {
          reject(err)
        }
      })
    })
  })

  it('loads items onFocus if not yet loaded', async () => {
    const children = ({ onFocus }) => <input onFocus={onFocus} />
    const loadItems = jest.fn()

    const component = renderComponent({
      children,
      loadItems,
    })

    component.find('input').simulate('focus')

    expect(loadItems).toHaveBeenCalled()
    loadItems.mockClear()

    component.find('input').simulate('focus')
    expect(loadItems).not.toHaveBeenCalled()
  })
})
