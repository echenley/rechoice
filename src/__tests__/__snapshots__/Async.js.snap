// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async renders function children prop 1`] = `
<div>
  {"items":[],"loadingState":"success"}
</div>
`;
